designs/v1alpha4-api.md

```diff
@@ -0,0 +1,130 @@
+
+```bash
+apiVersion: karpenter.sh/v1alpha4
+kind: Provisioner
+spec:
+  provider:
+    apiVersion: extensions.karpenter.sh/v1alpha1
+    kind: AWS
+    securityGroups: ["abc", "def"]
+    subnets: ["123", "456"]
+    launchTemplateName: "foo"
+```
+
+The Cloud Provider *may* continue to support corresponding well known labels at the pod level (e.g. `node.k8s.aws/subnet-name`).
+
+```bash
+# PodSpec: Simple key value constraints
+spec:
+  nodeSelector:
+    node.k8s.aws/subnet-name: "123" # Vendor Specific Field
+    kubernetes.io/instance-type: "m5.large" # Vendor Neutral Field
+```
+
+```bash
+# PodSpec: More expressive than node selectors, can specify multiple or preferences
+spec:
+  affinity:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+        - matchExpressions:
+          - key: node.k8s.aws/subnet-name
+            operator: In
+            values: ["123", "456"]
+```
+
+*Note: Not all `spec.provider` fields must have a corresponding well known label.*
+
+### Alternative: Inline vs Nested Parameters
+It's possible to leverage inline json capabilities to collapse these parameters to the top level. For example
+```bash
+spec:
+  zones: ["us-west-2a", "us-west-2b"]
+  securityGroups: ["abc", "def"]
+```
+
+1. (+) Parameters that are natural siblings (e.g. `zones`, `securityGroups`) will be no longer be separated by `provider`.
+2. (-) It's not immediately clear to readers which parameters are vendor specific.
+3. (-) Key name conflicts may arise between provider and vendor fields.
+4. (-) `kind` and `apiVersion` are awkwardly placed at the top level of spec.
+
+### Alternative: Duck Typing
+
+It's possible to follow [Knative's Duck Typing approach](https://www.youtube.com/watch?v=kldVg63Utuw) and build vendor specific CRDs that contain vendor neutral snippets (or ducks) that Karpenter can recognize. Cloud Providers would consume Karpenter generic controllers as libaries, which would behave against the generic API snippets. For example:
+
+```bash
+apiVersion: karpenter.k8s.aws/v1alpha1 # Vendor Specific
+kind: Provisioner
+spec:
+  limits: {} # Vendor neutral, recognized by Karpenter generic controllers
+  labels: {} # Vendor neutral, recognized by Karpenter generic controllers
+  taints: [] # Vendor netural, recognized by Karpenter generic controllers
+  subnets: [] # Vendor specific, only recognized by AWS Cloud Provider code
+```
+
+1. (+) Versioning is only defined in one place
+2. (+) Vendors may change versions decoupled from each other.
+3. (+) Parameters that are natural siblings (e.g. `zones`, `securityGroups`) will be no longer be separated by `provider`.
+4. (-) Cloud Provider must do more than simply implementing an interface (they must define an CR, etc).
+
+This approach is quite attractive, but requires significant effort and changes to the project.
```

Comment:
I'd probably re-word this a bit, as "quite attractive" is somewhat difficult to quantify. I'd suggest more focusing on how much effort it would be (perhaps quantify that a bit) as a reason not to do it. (I'm also not 100% sure all of the potential downsides were listed?)